src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+  }
+
+  std::vector<std::shared_ptr<const Table>> results(_output_row_count);
+
+  for (auto chunk_offset = ChunkOffset{0}; chunk_offset < _output_row_count; ++chunk_offset) {
+    results[chunk_offset] = _evaluate_select_expression_for_row(expression, chunk_offset);
+  }
+
+  return results;
+}
+
+std::shared_ptr<const Table> ExpressionEvaluator::_evaluate_select_expression_for_row(
+    const PQPSelectExpression& expression, const ChunkOffset chunk_offset) {
+  Assert(expression.parameters.empty() || _chunk,
+         "Sub-SELECT references external Columns but Expression doesn't operate on a Table/Chunk");
+
+  std::unordered_map<ParameterID, AllTypeVariant> parameters;
+
+  for (auto parameter_idx = size_t{0}; parameter_idx < expression.parameters.size(); ++parameter_idx) {
+    const auto& parameter_id_column_id = expression.parameters[parameter_idx];
+    const auto parameter_id = parameter_id_column_id.first;
+    const auto column_id = parameter_id_column_id.second;
+    const auto& column = *_chunk->get_column(column_id);
+
+    resolve_data_type(column.data_type(), [&](const auto data_type_t) {
+      using ColumnDataType = typename decltype(data_type_t)::type;
+
+      const auto column_materialization =
+          std::dynamic_pointer_cast<ExpressionResult<ColumnDataType>>(_column_materializations[column_id]);
+
+      if (column_materialization->null(chunk_offset)) {
+        parameters.emplace(parameter_id, NullValue{});
+      } else {
+        parameters.emplace(parameter_id, column_materialization->value(chunk_offset));
+      }
+    });
+  }
+
+  // TODO(moritz) recreate() shouldn't be necessary for every row if we could re-execute PQPs...
+  auto row_pqp = expression.pqp->recreate();
+  row_pqp->set_parameters(parameters);
+
+  SQLQueryPlan query_plan;
+  query_plan.add_tree_by_root(row_pqp);
+  const auto tasks = query_plan.create_tasks();
+  CurrentScheduler::schedule_and_wait_for_tasks(tasks);
+
+  return row_pqp->get_output();
+}
+
+std::shared_ptr<BaseColumn> ExpressionEvaluator::evaluate_expression_to_column(const AbstractExpression& expression) {
+  std::shared_ptr<BaseColumn> column;
+
+  _resolve_to_expression_result_view(expression, [&](const auto& view) {
+    using ColumnDataType = typename std::decay_t<decltype(view)>::Type;
+
+    // clang-format off
+    if constexpr (std::is_same_v<ColumnDataType, NullValue>) {
+      Fail("Can't create a Column from a NULL");
+    } else {
+      pmr_concurrent_vector<ColumnDataType> values(_output_row_count);
+
+      for (auto chunk_offset = ChunkOffset{0}; chunk_offset < _output_row_count; ++chunk_offset) {
+        values[chunk_offset] = view.value(chunk_offset);
+      }
+
+      if (view.is_nullable()) {
+        pmr_concurrent_vector<bool> nulls(_output_row_count);
+        for (auto chunk_offset = ChunkOffset{0}; chunk_offset < _output_row_count; ++chunk_offset) {
+          nulls[chunk_offset] = view.null(chunk_offset);
```

Comment:
This is one virtual method call per row, correct? Is it possible to resolve that once per view?

Reply:
Nope, there are _no virtual function calls_ involved here. `view.null(chunk_offset)` will resolve to a literal `false` if the view is non-nullable and to a vector-offset access otherwise.